### PR TITLE
fix(financial-aid): Redirect to status page on web-osk upgraded

### PR DESCRIPTION
**Pull request by MargretFinnboga**
# Redirect to status page on web-osk upgraded

https://app.asana.com/0/1200826552987384/1200368932625373

## What

Added return url enum to user 
Get application added to status page

## Why

Better login flow for web-osk

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/financial-aid/shared/src/lib/types.ts [hunk 1]
```diff
@@ -27,6 +27,12 @@ export enum RolesRule {
   VEITA = 'veita',
 }
 
+export enum ReturnUrl {
+  UMSOKN = '/umsokn',
+  STADA = '/stada',
+  VEITA = '/nymal',
```
Comment (Rikkiabb) [inline libs/financial-aid/shared/src/lib/types.ts:33]: Should we change to something like:
```
 APPLICATION = '/umsokn',
 MYPAGE = '/stada',
 ADMIN = '/nymal',
```
?

File: apps/financial-aid/api/src/app/modules/auth/auth.controller.ts [hunk 2]
```diff
@@ -172,7 +174,17 @@ export class AuthController {
...
+      user.returnUrl = ReturnUrl.VEITA
+    }
+    if (res.req?.query.applicationId) {
+      user.returnUrl = ReturnUrl.STADA
+    }
+
+    const returnURl =
+      user.returnUrl === ReturnUrl.STADA
+        ? ReturnUrl.STADA + '/' + res.req?.query.applicationId
+        : user.returnUrl
```
Comment (Rikkiabb) [inline apps/financial-aid/api/src/app/modules/auth/auth.controller.ts:187]: You could add the applicationId param on line 181 and then you don't have to have this check here.

File: apps/financial-aid/web-osk/graphql/sharedGql.ts [hunk 3]
```diff
@@ -62,3 +62,16 @@ export const CurrentUserQuery = gql`
     }
   }
 `
+
+export const GetApplicationQuery = gql`
+  query GetApplicantyQuery($input: ApplicationInput!) {
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/graphql/sharedGql.ts:67]: `GetApplicantyQuery` 🤔 should this be `GetApplicationQuery`

File: apps/financial-aid/web-osk/src/components/Login/login.tsx [hunk 4]
```diff
@@ -11,53 +12,59 @@ interface Props {
 const Login = ({ headline, about }: Props) => {
   const router = useRouter()
 
-  const apiLoginRoute = '/api/auth/login'
+  const apiLoginRoute = router.query.id
+    ? `/api/auth/login?applicationId=${router.query.id}&nationalId=`
+    : '/api/auth/login?nationalId='
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/src/components/Login/login.tsx:17]: Should we concat `nationald=` where use them below?

And should we have these routes as const in this file?

File: apps/financial-aid/web-osk/src/components/Status/InProgress.tsx [hunk 5]
```diff
@@ -41,7 +41,7 @@ const InProgress = ({ currentApplication }: Props) => {
             cta={{
               label: 'Hlaða upp gögnum',
               onClick: () => {
-                router.push(`${router.query.id}/gogn`)
+                router.push(`stada/${router.query.id}/gogn`)
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/src/components/Status/InProgress.tsx:44]: Should we maybe have all routes in a const file? Then in the future if we add something like `stada` then we only change that file.

File: apps/financial-aid/web-osk/src/routes/status/MainPage/mainPage.tsx [hunk 6]
```diff
@@ -15,22 +16,40 @@ import {
...
+      variables: { input: { id: router.query.id } },
+      fetchPolicy: 'no-cache',
+      errorPolicy: 'all',
+    },
+  )
   const currentApplication = useMemo(() => {
-    if (user?.currentApplication) {
-      return user.currentApplication
+    if (data?.application) {
+      return data?.application
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/src/routes/status/MainPage/mainPage.tsx:50]: is the question mark needed on line 50?

**Review by Rikkiabb — CHANGES_REQUESTED** [review]
🔴

File: apps/financial-aid/web-osk/src/components/Login/login.tsx (hunk 4 above)
Comment (MargretFinnboga) [reply apps/financial-aid/web-osk/src/components/Login/login.tsx]: Changed to two different ones for the time being, while we are still using fake users. What about that?

File: apps/financial-aid/api/src/app/modules/auth/auth.controller.ts (hunk 2 above)
Comment (MargretFinnboga) [reply apps/financial-aid/api/src/app/modules/auth/auth.controller.ts:187]: true, but I cant set the user.returnurl with application id it only accepts enum, so I will always have to add another variable anyways?

Comment (Rikkiabb) [reply apps/financial-aid/api/src/app/modules/auth/auth.controller.ts:187]: So what I am thinking is basically you could drop lines 177-187 and instead do something like: 
```
        const returnUrl = `${user.returnUrl}${
          res.req?.query.applicationId ? `/${res.req.query.applicationId}` : ''
        }`
```
Or you could even do it in the redirect on line 198 or am I misunderstanding something? 👼

File: apps/financial-aid/web-osk/src/components/Login/login.tsx [hunk 7]
```diff
@@ -11,53 +12,63 @@ interface Props {
 const Login = ({ headline, about }: Props) => {
   const router = useRouter()
 
-  const apiLoginRoute = '/api/auth/login'
+  const apiLoginRouteForFake = router.query.id
+    ? `/api/auth/login?applicationId=${router.query.id}&nationalId=`
+    : '/api/auth/login?'
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/src/components/Login/login.tsx:17]: Shouldn't line 17 also include `nationalId=`?

File: apps/financial-aid/web-osk/src/routes/status/MainPage/mainPage.tsx [hunk 8]
```diff
@@ -16,22 +17,40 @@ import {
...
+import {
+  Application,
+  CurrentApplication,
+  getActiveTypeForStatus,
+} from '@island.is/financial-aid/shared'
+
 import { useLogOut } from '@island.is/financial-aid-web/osk/src/utils/useLogOut'
+import { useQuery } from '@apollo/client'
+import { useRouter } from 'next/router'
+import { GetApplicationQuery } from '@island.is/financial-aid-web/oskgraphql'
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/src/routes/status/MainPage/mainPage.tsx:31]: Some imports here not used

File: libs/financial-aid/shared/src/lib/const.ts [hunk 9]
```diff
@@ -11,3 +11,5 @@ export const AllowedFakeUsers = [
   '0000000002', // Veita user
   '0000000003', // User with an application which needs data
 ]
+
+export const LoginStatusPage = '/stada'
```
Comment (Rikkiabb) [inline libs/financial-aid/shared/src/lib/const.ts:15]: So what I was thinking here is something like this: 

```
export const Routes = {
  status: '/stada',
  statusFileUpload: (id: string) => `/stada/${id}/gogn`,
}
```

Etc....

Then we would never have hardcoded values in our components but would only refer to the `Routes` obj. Does this make sense? It could also be a separate PR.

**Review by Rikkiabb — COMMENTED** [review]
Awesome, couple of comments + what we talked about IRL

File: apps/financial-aid/api/src/app/modules/auth/auth.controller.ts [hunk 10]
```diff
@@ -172,12 +174,26 @@ export class AuthController {
...
+  getReturnUrl = (service: string, applicationId: string) => {
+    switch (true) {
+      case service === 'veita':
+        return ReturnUrl.ADMIN
+      case applicationId != undefined:
+        return `${ReturnUrl.MYPAGE}/${applicationId}`
+      default:
+        return ReturnUrl.APPLICATION
+    }
+  }
```
Comment (Rikkiabb) [inline apps/financial-aid/api/src/app/modules/auth/auth.controller.ts:186]: How do you feel about?
```
  getReturnUrl = (service: RolesRule, applicationId: string | undefined) => {
    switch (true) {
      case service === RolesRule.OSK && applicationId !== undefined:
        return `${ReturnUrl.MYPAGE}/${applicationId}`
      case service === RolesRule.VEITA:
        return ReturnUrl.ADMIN
      default:
        return ReturnUrl.APPLICATION
    }
  }
```

File: apps/financial-aid/web-osk/src/components/HasApplied/hasApplied.tsx [hunk 11]
```diff
@@ -18,7 +19,7 @@ const HasApplied = () => {
 
       <Button
         onClick={() => {
-          router.push(`/${user?.currentApplication?.id}`)
+          router.push(`${Routes.status}/${user?.currentApplication?.id}`)
```
Comment (Rikkiabb) [inline apps/financial-aid/web-osk/src/components/HasApplied/hasApplied.tsx:22]: This could be in the Routes object instead of interpolating

**Review by Rikkiabb — APPROVED** [review]
Some comments, you can look at but feel free to merge
